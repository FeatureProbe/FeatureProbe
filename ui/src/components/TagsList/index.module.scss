.tags-list {
  display: flex;
}

.tags {
  background: #ececf6;
  border-radius: 4px;
  font-size: 10px;
  color: #495057;
  padding: 0 4px;
  max-width: 97px;
  height: 16px;
  margin: 4px 2px;
}

.tags-text {
  width: 120%;
  transform: scale(0.83, 0.83);
  font-size: 12px;
  display: block;
  transform-origin: left center;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  line-height: 16px;
}

.tags-more {
  font-size: 12px;
  color: #495057;
  height: 16px;
  margin-top: 2px;
  margin-left: 2px;
  display: inline-block;
  cursor: pointer;
}

.tags-more:hover {
  color: #556ee6;
}

.popup-list {
  width: 260px;
  display: flex;
  flex-wrap: wrap;
}