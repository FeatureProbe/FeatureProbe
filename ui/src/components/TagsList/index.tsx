import { SyntheticEvent, useEffect, useState } from 'react';
import { Popup } from 'semantic-ui-react';
import styles from './index.module.scss';

interface IProps {
  tags: string[];
  showCount?: number;
}

const TagsList = (props: IProps) => {
  const { tags, showCount } = props;
  const [ showTags, saveShowTags ] = useState<string[]>([]);
  const [ popupTags, savePopupTags ] = useState<string[]>([]);
  
  useEffect(() => {
    const length = tags.length;
    const maxShowCount = showCount || 5;

    if(length > maxShowCount) {
      saveShowTags(tags.slice(0, maxShowCount));
      savePopupTags(tags.slice(maxShowCount));
    } else {
      saveShowTags(tags);
    }
  }, [tags, showCount]);

  return (
    <div className={styles['tags-list']}>
      {
        showTags.map((val: string) => {
          return (
            <div className={styles.tags} key={val}>
              <span className={styles['tags-text']}>
                { val }
              </span>
            </div>
          );
        })
      }

      {
        popupTags.length > 0 && (
          <Popup
            basic
            on='click'
            position='bottom center'
            className={styles.popup}
            trigger={
              <span 
                className={styles['tags-more']} 
                onClick={(e: SyntheticEvent) => {
                  document.body.click();
                  e.stopPropagation();
                }}
              > 
                +{ popupTags.length }...
              </span>
            }
          >
            <div className={styles['popup-list']}>
              {
                popupTags.map((val: string) => {
                  return (
                    <div className={styles.tags} key={val}>
                      <span className={styles['tags-text']}>
                        { val }
                      </span>
                    </div>
                  );
                })
              }
            </div>
          </Popup>
        )
      }
    </div>
  );
};
  
export default TagsList;