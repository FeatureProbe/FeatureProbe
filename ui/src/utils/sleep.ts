const sleep = (delay: number) => {
  return new Promise((res) => {
    setTimeout(() => {
      res('ok');
    }, delay);
  });
};

export default sleep;
