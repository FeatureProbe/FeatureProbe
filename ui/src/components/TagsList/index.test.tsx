import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import TagsList from '.';
import sleep from 'utils/sleep';

it('TagsList snapshot', (done) => {
  (async () => {
    const { asFragment } = render(<TagsList tags={['test1', 'test2']} />);
    expect(asFragment()).toMatchSnapshot();
    done();
  })();
});

it('click TagsList more', (done) => {
    (async () => {
      const { baseElement } = render(<TagsList tags={['test1', 'test2']} showCount={1} />);
      expect(baseElement).toMatchSnapshot();

      await userEvent.click(screen.getByText('+1...'));
      await sleep(1000);

      expect(baseElement).toMatchSnapshot();
      done();
    })();
  });