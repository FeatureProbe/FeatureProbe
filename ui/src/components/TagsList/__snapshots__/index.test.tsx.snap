// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsList snapshot 1`] = `
<DocumentFragment>
  <div
    class="tags-list"
  >
    <div
      class="tags"
    >
      <span
        class="tags-text"
      >
        test1
      </span>
    </div>
    <div
      class="tags"
    >
      <span
        class="tags-text"
      >
        test2
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`click TagsList more 1`] = `
<body>
  <div>
    <div
      class="tags-list"
    >
      <div
        class="tags"
      >
        <span
          class="tags-text"
        >
          test1
        </span>
      </div>
      <span
        class="tags-more"
      >
        +
        1
        ...
      </span>
    </div>
  </div>
</body>
`;

exports[`click TagsList more 2`] = `
<body>
  <div>
    <div
      class="tags-list"
    >
      <div
        class="tags"
      >
        <span
          class="tags-text"
        >
          test1
        </span>
      </div>
      <span
        class="tags-more"
      >
        +
        1
        ...
      </span>
    </div>
  </div>
  <div
    style="display: flex; position: absolute; left: 0px; top: 0px; transform: translate(0px, 0px);"
  >
    <div
      class="ui bottom center basic popup transition visible popup"
      style="position: initial;"
    >
      <div
        class="popup-list"
      >
        <div
          class="tags"
        >
          <span
            class="tags-text"
          >
            test2
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;
